Add r-gdalraster

ctoney opened the pull request:
* Add `gdalraster` v2.1.0.
  * c.f. https://github.com/USDAForestService/gdalraster
* c.f. https://github.com/USDAForestService/gdalraster/issues/758
* Used the helper script: https://github.com/bgruening/conda_r_skeleton_helper

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- NA If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/r-gdalraster/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/r-gdalraster/meta.yaml**:

 * ❌ This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.

For **recipes/r-gdalraster/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/16530257011](https://github.com/conda-forge/conda-forge-webservices/actions/runs/16530257011). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/r-gdalraster/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipes/r-gdalraster/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/16530484268](https://github.com/conda-forge/conda-forge-webservices/actions/runs/16530484268). Examine the logs at this URL for more detail.</sub>

ctoney commented:
Started with the recipe generated by the helper script: https://github.com/bgruening/conda_r_skeleton_helper.

Added under `requirements: build:`
```
    - {{ stdlib("c") }}
```

and

```
    - r-rcpp
    - r-rcppint64
    - r-bit64
    - r-nanoarrow
    - r-wk
    - r-xml2
```

Added under `requirements: host:`
```
    - libgdal-core
```

Windows builds currently fail with
```
x86_64-w64-mingw32-g++ -shared -s -o gdalraster.dll tmp.def RcppExports.o cmb_table.o gdal_dt.o gdal_exp.o gdal_vsi.o gdalraster.o gdalvector.o geom_api.o ogr_util.o progress_r.o rcpp_util.o running_stats.o srs_api.o transform.o vsifile.o -fopenmp -lgdal -larmadillo -lopenblas -lgfortran -lquadmath -lpq -lpgcommon -lpgport -lodbc32 -lodbccp32 -lblosc -lkea -lhdf5_cpp -lhdf5 -lpoppler -llcms2 -lfreetype -lharfbuzz -lfreetype -llz4 -lpcre2-8 -lxml2 -lopenjp2 -lnetcdf -lmysqlclient -lspatialite -lgeos_c -lgeos -lminizip -lgeos -ljson-c -lgta -lfreexl -lexpat -lssl -lpsapi -lgif -lmfhdf -lhdf5_hl -lcrypto -lportablexdr -ldf -lhdf5 -lsz -lpng16 -lpng -lpoppler -llcms2 -lfreetype -lharfbuzz -lfreetype -llz4 -lpcre2-8 -lpcre -lcurl -lbcrypt -lrtmp -lssl -lssh2 -lidn2 -lunistring -liconv -lgcrypt -lcrypto -lgpg-error -lws2_32 -ltiff -llzma -ljpeg -lz -lcfitsio -lzstd -lwebpdecoder -lwebp -lsbml-static -lgeotiff -lproj -lsqlite3 -lbz2 -lcrypt32 -lwldap32 -lsecur32 -LD:/bld/r-gdalraster_1753479326626/_h_env/Library/lib/x64 -LD:/bld/r-gdalraster_1753479326626/_h_env/Library/lib -LD:/bld/r-gdalraster_1753479326626/_h_env/lib/R/bin/x64 -lR
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -larmadillo: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lopenblas: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lpq: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lpgcommon: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lpgport: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lkea: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lhdf5_cpp: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lhdf5: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lpoppler: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -llcms2: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lcfitsio: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lsbml-static: No such file or directory
D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lgeotiff: No such file or directory
collect2.exe: error: ld returned 1 exit status
```

matthewfeickert commented on recipes/r-gdalraster/meta.yaml:
```diff
@@ -0,0 +1,134 @@
+  recipe-maintainers:
+    - conda-forge/r
+    - ctoney
+
+# Package: gdalraster
+# Title: Bindings to 'GDAL'
+# Version: 2.1.0
+# Authors@R: c( person("Chris", "Toney", email = "jctoney@gmail.com", role = c("aut", "cre"), comment = "R interface/additional functionality"), person("Michael D.","Sumner", role = c("ctb")), person("Frank", "Warmerdam", role = c("ctb", "cph"), comment = "GDAL API documentation; src/progress_r.cpp from gdal/port/cpl_progress.cpp"), person("Even", "Rouault", role = c("ctb", "cph"), comment = "GDA ...
+# Description: API bindings to the Geospatial Data Abstraction Library ('GDAL', <https://gdal.org>). Implements the 'GDAL' Raster and Vector Data Models. Bindings are implemented with 'Rcpp' modules. Exposed C++ classes and stand-alone functions wrap much of the 'GDAL' API and provide additional functionality. Calling signatures resemble the native C, C++ and Python APIs provided by the 'GDAL' pr ...
+# License: MIT + file LICENSE
+# Copyright: See file inst/COPYRIGHTS for details.
+# URL: https://usdaforestservice.github.io/gdalraster/, https://github.com/USDAForestService/gdalraster
+# BugReports: https://github.com/USDAForestService/gdalraster/issues
+# Depends: R (>= 4.2.0)
+# Imports: bit64, graphics, grDevices, methods, nanoarrow, Rcpp (>= 1.0.7), stats, tools, utils, wk, xml2
+# LinkingTo: nanoarrow, Rcpp, RcppInt64
+# Suggests: gt, knitr, rmarkdown, scales, testthat (>= 3.0.0)
+# NeedsCompilation: yes
+# SystemRequirements: C++17, GDAL (>= 3.1.0, built against GEOS)
+# Encoding: UTF-8
+# RoxygenNote: 7.3.2
+# VignetteBuilder: knitr
+# Config/testthat/edition: 3
+# Packaged: 2025-07-11 23:34:52 UTC; ctoney
+# Author: Chris Toney [aut, cre] (R interface/additional functionality), Michael D. Sumner [ctb], Frank Warmerdam [ctb, cph] (GDAL API documentation; src/progress_r.cpp from gdal/port/cpl_progress.cpp), Even Rouault [ctb, cph] (GDAL API documentation), Marius Appel [ctb, cph] (configure.ac based on https://github.com/appelmar/gdalcubes), Daniel James [ctb, cph] (Boost combine hashes method in src ...
+# Maintainer: Chris Toney <jctoney@gmail.com>
+# Repository: CRAN
+# Date/Publication: 2025-07-14 19:30:03 UTC
```
```suggestion
```

This is all cruft that shouldn't exist, but hasn't gotten removed for some reason from the recipe generator.

matthewfeickert commented on recipes/r-gdalraster/meta.yaml:
```diff
@@ -0,0 +1,134 @@
+    - r-nanoarrow
+    - r-wk
+    - r-xml2
+    - {{ compiler('c') }}              # [not win]
+    - {{ stdlib("c") }}                # [not win]
+    - {{ compiler('m2w64_c') }}        # [win]
+    - {{ stdlib("c") }}                # [win]
+    - {{ compiler('cxx') }}            # [not win]
+    - {{ compiler('m2w64_cxx') }}      # [win]
```
```suggestion
    - {{ stdlib("c") }}
    - {{ compiler("cxx") }}
```

Unless there is something very weird with R on conda-forge compared to every other language this can just use the the standard compiler metapackages which know how to resolve correctly.

The CXX compilers should also get the required C compilers, so there shouldn't be a need for both, AFAIK.

cc @conda-forge/r for clarification on why the recipe generator is recommending this.

matthewfeickert commented on recipes/r-gdalraster/meta.yaml:
```diff
@@ -0,0 +1,134 @@
+    - r-wk
+    - r-xml2
+
+test:
+  commands:
+    - $R -e "library('gdalraster')"           # [not win]
+    - "\"%R%\" -e \"library('gdalraster')\""  # [win]
```
```suggestion
test:
  requires:
    - r-testthat
  commands:
    - $R -e "library('gdalraster')"  # [not win]
    - $R -e "testthat::test_file('tests/testthat.R', stop_on_failure=TRUE)"  # [not win]
    - "\"%R%\" -e \"library('gdalraster')\""  # [win] 
    - "\"%R%\" -e \"testthat::test_file('tests/testthat.R', stop_on_failure=TRUE)\""  # [win]
```

If `gdalraster` has a R test suite it should get run here too (also anything beyond `testthat` should get added. If not, ignore this, but consider adding any existing test suites a different way.

matthewfeickert reviewed (commented):
Haven't looked at anything in depth, just a pass in the airport.

matthewfeickert commented:
> Windows builds currently fail with
> 
> ```
> x86_64-w64-mingw32-g++ -shared -s -o gdalraster.dll tmp.def RcppExports.o cmb_table.o gdal_dt.o gdal_exp.o gdal_vsi.o gdalraster.o gdalvector.o geom_api.o ogr_util.o progress_r.o rcpp_util.o running_stats.o srs_api.o transform.o vsifile.o -fopenmp -lgdal -larmadillo -lopenblas -lgfortran -lquadmath -lpq -lpgcommon -lpgport -lodbc32 -lodbccp32 -lblosc -lkea -lhdf5_cpp -lhdf5 -lpoppler -llcms2 -lfreetype -lharfbuzz -lfreetype -llz4 -lpcre2-8 -lxml2 -lopenjp2 -lnetcdf -lmysqlclient -lspatialite -lgeos_c -lgeos -lminizip -lgeos -ljson-c -lgta -lfreexl -lexpat -lssl -lpsapi -lgif -lmfhdf -lhdf5_hl -lcrypto -lportablexdr -ldf -lhdf5 -lsz -lpng16 -lpng -lpoppler -llcms2 -lfreetype -lharfbuzz -lfreetype -llz4 -lpcre2-8 -lpcre -lcurl -lbcrypt -lrtmp -lssl -lssh2 -lidn2 -lunistring -liconv -lgcrypt -lcrypto -lgpg-error -lws2_32 -ltiff -llzma -ljpeg -lz -lcfitsio -lzstd -lwebpdecoder -lwebp -lsbml-static -lgeotiff -lproj -lsqlite3 -lbz2 -lcrypt32 -lwldap32 -lsecur32 -LD:/bld/r-gdalraster_1753479326626/_h_env/Library/lib/x64 -LD:/bld/r-gdalraster_1753479326626/_h_env/Library/lib -LD:/bld/r-gdalraster_1753479326626/_h_env/lib/R/bin/x64 -lR
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -larmadillo: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lopenblas: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lpq: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lpgcommon: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lpgport: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lkea: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lhdf5_cpp: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lhdf5: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lpoppler: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -llcms2: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lcfitsio: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lsbml-static: No such file or directory
> D:/bld/r-gdalraster_1753479326626/_h_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/15.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lgeotiff: No such file or directory
> collect2.exe: error: ld returned 1 exit status
> ```

These failing linker flags don't appear in the Linux and macOS builds. If Windows builds need these libraries then they need to be added as `host` requirements with the `# [win]` selector.

matthewfeickert commented on recipes/r-gdalraster/meta.yaml:
```diff
@@ -0,0 +1,113 @@
+    - {{ cran_mirror }}/src/contrib/Archive/gdalraster/gdalraster_{{ version }}.tar.gz
+  sha256: 5a52e24e10e44163dfc366007a356dfc27f9d28353d557b507e46d119f3b6ef2
+  patches:
+    - min-ucrt-libs.patch
```
```suggestion
    - min-ucrt-libs.patch  # [win]
```

If this patch is only needed for Windows can just apply it as such. Though is fine to keep it generic also.

matthewfeickert commented:
(I can take another look at this on Sunday US time or Monday, but I realistically might not be able to give _that_ much advice on Windows builds as I don't have access to a Windows machine beyond spamming GitHub Actions runners.)

matthewfeickert commented:
> I noticed that the recipes for two other R packages with GDAL dependency (`gdalcubes` and `terra`) have
> 
> ```
> build:
>   skip: true  # [win]
> ```
> 
> I assume that means Windows build is currently skipped?

Yes.

@ctoney, my personal recommendation would be that if you can get things building for Linux and macOS to for now skip Windows just to get `r-gdalraster` on conda-forge. Once it exists as a feedstock you can do additional debugging to try to get Windows builds working.


> Both packages show win-64 installers available (e.g., https://anaconda.org/conda-forge/r-gdalcubes), but versions on win-64 are older compared with installers for the the other platforms.

Yeah, it seems that something with Windows builds got annoying enough to shut them off. @conda-forge/r-gdalcubes @conda-forge/r-terra @conda-forge/r-sf @conda-forge/r can any of you give context on the Windows issues or what is needed to work around them?

danielnachun reviewed (approved):
While I was able to get `r-sf` to build with `r_clang`, `r-terra` has not worked so far.  Let's get this feedstock made and get the cross-compilation working on macOS and Linux and then we can open a PR to wrestle with Windows afterwards.
